File: iosApp/iosApp/Features/ModelFileBrowser/ModelFileBrowserScreen.swift
import SwiftUI
import Shared

struct ModelFileBrowserScreen: View {
    @StateObject private var viewModel = ModelFileBrowserViewModel()
    @State private var showAddDialog = false
    @State private var newDirectoryPath = ""

    var body: some View {
        List {
            directoriesSection
            if !viewModel.files.isEmpty {
                filesSection
            }
        }
        .navigationTitle("Model Files")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar { toolbarContent }
        .task { await viewModel.observeDirectories() }
        .task { await viewModel.observeFiles() }
        .overlay { emptyOverlay }
        .alert("Add Model Directory", isPresented: $showAddDialog) {
            addDirectoryAlert
        }
        .alert("Scan Error", isPresented: .init(
            get: { viewModel.errorMessage != nil },
            set: { if !$0 { viewModel.onDismissError() } }
        )) {
            Button("OK") { viewModel.onDismissError() }
        } message: {
            Text(viewModel.errorMessage ?? "")
        }
    }

    // MARK: - Sections

    private var directoriesSection: some View {
        Section("Directories") {
            ForEach(viewModel.directories, id: \.id) { directory in
                DirectoryRow(directory: directory) {
                    viewModel.onRemoveDirectory(directory.id)
                }
            }
            Button {
                showAddDialog = true
            } label: {
                Label("Add Directory", systemImage: "plus")
            }
        }
    }

    private var filesSection: some View {
        Section("Scanned Files (\(viewModel.files.count))") {
            if viewModel.scanStatus == .scanning || viewModel.scanStatus == .verifying {
                HStack(spacing: Spacing.sm) {
                    ProgressView()
                    Text(viewModel.scanProgress)
                        .font(.civitBodySmall)
                        .foregroundColor(.civitOnSurfaceVariant)
                }
            }
            ForEach(viewModel.files, id: \.id) { file in
                ModelFileRow(file: file)
            }
        }
    }

    // MARK: - Toolbar

    @ToolbarContentBuilder
    private var toolbarContent: some ToolbarContent {
        ToolbarItem(placement: .topBarTrailing) {
            if viewModel.scanStatus == .scanning || viewModel.scanStatus == .verifying {
                ProgressView()
            } else if !viewModel.directories.isEmpty {
                Button {
                    viewModel.onScanAll()
                } label: {
                    Image(systemName: "arrow.clockwise")
                }
            }
        }
    }

    // MARK: - Empty State

    @ViewBuilder
    private var emptyOverlay: some View {
        if viewModel.directories.isEmpty {
            VStack(spacing: Spacing.sm) {
                Text("No model directories configured")
                    .font(.civitTitleMedium)
                Text("Tap + to add a directory path")
                    .font(.civitBodyMedium)
                    .foregroundColor(.civitOnSurfaceVariant)
            }
        }
    }

    // MARK: - Add Dialog

    @AlertActionsBuilder
    private var addDirectoryAlert: some View {
        TextField("/path/to/models", text: $newDirectoryPath)
        Button("Add") {
            viewModel.onAddDirectory(newDirectoryPath)
            newDirectoryPath = ""
        }
        Button("Cancel", role: .cancel) {
            newDirectoryPath = ""
        }
    }
}

// MARK: - Subviews

private struct DirectoryRow: View {
    let directory: ModelDirectory
    let onRemove: () -> Void
    @State private var showConfirm = false

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xs) {
            Text(directory.label ?? directory.path)
                .font(.civitBodyMedium)
                .lineLimit(1)
            Text(directory.path)
                .font(.civitBodySmall)
                .foregroundColor(.civitOnSurfaceVariant)
                .lineLimit(1)
        }
        .swipeActions(edge: .trailing) {
            Button(role: .destructive) {
                showConfirm = true
            } label: {
                Label("Remove", systemImage: "trash")
            }
        }
        .alert("Remove Directory", isPresented: $showConfirm) {
            Button("Remove", role: .destructive) { onRemove() }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("Remove this directory and all its scanned data?")
        }
    }
}

private struct ModelFileRow: View {
    let file: LocalModelFile

    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: Spacing.xs) {
                Text(file.fileName)
                    .font(.civitBodyMedium)
                    .lineLimit(1)
                Text(formatFileSize(file.sizeBytes))
                    .font(.civitBodySmall)
                    .foregroundColor(.civitOnSurfaceVariant)
                if let match = file.matchedModel {
                    Text("\(match.modelName) - \(match.versionName)")
                        .font(.civitBodySmall)
                        .foregroundColor(.civitPrimary)
                        .lineLimit(1)
                    if match.hasUpdate {
                        Label("Update available", systemImage: "exclamationmark.triangle")
                            .font(.civitLabelSmall)
                            .foregroundColor(.civitTertiary)
                    }
                }
            }
            Spacer()
            if file.matchedModel != nil {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundColor(.civitPrimary)
            }
        }
    }

    private func formatFileSize(_ bytes: Int64) -> String {
        let kb = Double(bytes) / 1024.0
        let mb = kb / 1024.0
        let gb = mb / 1024.0
        if gb >= 1.0 {
            return String(format: "%.1f GB", gb)
        } else if mb >= 1.0 {
            return String(format: "%.1f MB", mb)
        } else {
            return String(format: "%.0f KB", kb)
        }
    }
}

File: androidApp/src/main/kotlin/com/riox432/civitdeck/ui/modelfiles/ModelFileBrowserScreen.kt
package com.riox432.civitdeck.ui.modelfiles

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.riox432.civitdeck.R
import com.riox432.civitdeck.domain.model.LocalModelFile
import com.riox432.civitdeck.domain.model.ModelDirectory
import com.riox432.civitdeck.domain.model.ScanStatus
import com.riox432.civitdeck.ui.theme.Spacing

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ModelFileBrowserScreen(
    viewModel: ModelFileBrowserViewModel,
    onBack: () -> Unit,
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()
    var showAddDialog by rememberSaveable { mutableStateOf(false) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Model Files") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = stringResource(R.string.cd_navigate_back)
                        )
                    }
                },
                actions = {
                    if (state.directories.isNotEmpty()) {
                        ScanButton(state.scanStatus, viewModel::onScanAll)
                    }
                },
            )
        },
        floatingActionButton = {
            FloatingActionButton(onClick = { showAddDialog = true }) {
                Icon(Icons.Default.Add, contentDescription = stringResource(R.string.cd_add_directory))
            }
        },
    ) { padding ->
        Column(modifier = Modifier.fillMaxSize().padding(padding)) {
            ScanProgressBar(state.scanStatus, state.scanProgress)
            if (state.directories.isEmpty()) {
                EmptyState()
            } else {
                FileList(state.directories, state.files, viewModel::onRemoveDirectory)
            }
        }
    }

    if (showAddDialog) {
        AddDirectoryDialog(
            onAdd = { viewModel.onAddDirectory(it) },
            onDismiss = { showAddDialog = false },
        )
    }

    state.errorMessage?.let { error ->
        AlertDialog(
            onDismissRequest = viewModel::onDismissError,
            title = { Text("Scan Error") },
            text = { Text(error) },
            confirmButton = { TextButton(onClick = viewModel::onDismissError) { Text("OK") } },
        )
    }
}

@Composable
private fun ScanButton(status: ScanStatus, onScan: () -> Unit) {
    val isScanning = status == ScanStatus.Scanning || status == ScanStatus.Verifying
    IconButton(onClick = onScan, enabled = !isScanning) {
        if (isScanning) {
            CircularProgressIndicator(modifier = Modifier.size(24.dp), strokeWidth = 2.dp)
        } else {
            Icon(Icons.Default.Refresh, contentDescription = stringResource(R.string.cd_scan_all_directories))
        }
    }
}

@Composable
private fun ScanProgressBar(status: ScanStatus, progress: String) {
    if (status == ScanStatus.Scanning || status == ScanStatus.Verifying) {
        Column(modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.lg)) {
            LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
            Text(
                text = progress,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = Spacing.xs, bottom = Spacing.sm),
            )
        }
    }
}

@Composable
private fun EmptyState() {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(
                "No model directories configured",
                style = MaterialTheme.typography.titleMedium,
            )
            Text(
                "Tap + to add a directory path",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = Spacing.sm),
            )
        }
    }
}

@Composable
private fun FileList(
    directories: List<ModelDirectory>,
    files: List<LocalModelFile>,
    onRemoveDirectory: (Long) -> Unit,
) {
    LazyColumn(modifier = Modifier.fillMaxSize()) {
        item {
            Text(
                "Directories",
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier.padding(horizontal = Spacing.lg, vertical = Spacing.sm),
            )
        }
        items(directories, key = { it.id }) { directory ->
            DirectoryItem(directory, onRemoveDirectory)
        }
        if (files.isNotEmpty()) {
            item {
                Text(
                    "Scanned Files (${files.size})",
                    style = MaterialTheme.typography.titleSmall,
                    color = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.padding(
                        horizontal = Spacing.lg,
                        vertical = Spacing.sm,
                    ),
                )
            }
            items(files, key = { it.id }) { file ->
                ModelFileItem(file)
            }
        }
    }
}

@Composable
private fun DirectoryItem(directory: ModelDirectory, onRemove: (Long) -> Unit) {
    var showConfirm by remember { mutableStateOf(false) }
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.lg, vertical = Spacing.sm),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                directory.label ?: directory.path,
                style = MaterialTheme.typography.bodyLarge,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                directory.path,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
        IconButton(onClick = { showConfirm = true }) {
            Icon(
                Icons.Default.Delete,
                contentDescription = stringResource(R.string.cd_remove),
                tint = MaterialTheme.colorScheme.error
            )
        }
    }
    if (showConfirm) {
        AlertDialog(
            onDismissRequest = { showConfirm = false },
            title = { Text("Remove Directory") },
            text = { Text("Remove this directory and all its scanned data?") },
            confirmButton = {
                TextButton(onClick = {
                    onRemove(directory.id)
                    showConfirm = false
                }) { Text("Remove") }
            },
            dismissButton = {
                TextButton(onClick = { showConfirm = false }) { Text("Cancel") }
            },
        )
    }
}

@Composable
private fun ModelFileItem(file: LocalModelFile) {
    Card(
        modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.lg, vertical = Spacing.xs),
        colors = CardDefaults.cardColors(
            containerColor = if (file.matchedModel != null) {
                MaterialTheme.colorScheme.primaryContainer.copy(alpha = 0.3f)
            } else {
                MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.3f)
            },
        ),
    ) {
        Row(
            modifier = Modifier.padding(Spacing.md),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    file.fileName,
                    style = MaterialTheme.typography.bodyMedium,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    formatFileSize(file.sizeBytes),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                file.matchedModel?.let { MatchedModelInfo(it) }
            }
            file.matchedModel?.let {
                Icon(
                    Icons.Default.CheckCircle,
                    contentDescription = stringResource(R.string.cd_matched),
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(24.dp),
                )
            }
        }
    }
}

@Composable
private fun MatchedModelInfo(match: com.riox432.civitdeck.domain.model.MatchedModelInfo) {
    Text(
        "${match.modelName} - ${match.versionName}",
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.primary,
        maxLines = 1,
        overflow = TextOverflow.Ellipsis,
    )
    if (match.hasUpdate) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                Icons.Default.Warning,
                contentDescription = stringResource(R.string.cd_update_available),
                modifier = Modifier.size(14.dp),
                tint = MaterialTheme.colorScheme.tertiary,
            )
            Text(
                "Update available",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.tertiary,
                modifier = Modifier.padding(start = Spacing.xs),
            )
        }
    }
}

@Composable
private fun AddDirectoryDialog(onAdd: (String) -> Unit, onDismiss: () -> Unit) {
    var path by rememberSaveable { mutableStateOf("") }
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Add Model Directory") },
        text = {
            OutlinedTextField(
                value = path,
                onValueChange = { path = it },
                placeholder = { Text("/path/to/models") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
        },
        confirmButton = {
            TextButton(
                onClick = {
                    onAdd(path)
                    onDismiss()
                },
                enabled = path.isNotBlank(),
            ) { Text("Add") }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Cancel") }
        },
    )
}

private fun formatFileSize(bytes: Long): String {
    val kb = bytes / 1024.0
    val mb = kb / 1024.0
    val gb = mb / 1024.0
    return when {
        gb >= 1.0 -> "%.1f GB".format(gb)
        mb >= 1.0 -> "%.1f MB".format(mb)
        else -> "%.0f KB".format(kb)
    }
}
